### ISSUE TITLE
Memory leak during backprop() in PyTorch 1.0.0

### ISSUE THREAD

**Issue by peterszabo77** [pytorch/pytorch#15799]
## 🐛 Bug

The RAM consumption continuously grows with each backpropagation step.
It can be easily checked using the official Reinforcement Learning tutorial:
[https://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html](https://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html)

## To Reproduce

1. Download the official reinforcement learning tutorial (python file):
[https://pytorch.org/tutorials/_downloads/b8954cc7b372cac10a92b8c6183846a3/reinforcement_q_learning.py](https://pytorch.org/tutorials/_downloads/b8954cc7b372cac10a92b8c6183846a3/reinforcement_q_learning.py)
2. Set num_episodes to a large value (e.g. 5000) to see the increasing memory consumption
3. Set device to cpu (I had only cpu available, but maybe the same happens with gpu)

## Expected behavior

Memory (RAM) consumption gradually increases.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).
 - PyTorch Version: 1.0.0
 - OS: Linux
 - How you installed PyTorch: pip
 - Build command you used (if compiling from source):
 - Python version: 3.5.3
 - CUDA/cuDNN version: no CUDA
 - GPU models and configuration: no GPU
 - Any other relevant information:

cc @ezyang @gchanan @SsnL @albanD

**Comment by zou3519** [comment]
@peterszabo77 do you ever OOM due to high memory usage? The memory allocator might just not be freeing its memory unless there is memory pressure.

**Comment by peterszabo77** [comment]
I did not check OOM on this particular code only the increasing usage, but probably yes. I did it because I wanted to check my similar experience in my custom, but much larger, reinforcement learning model with similar structure on a much simpler and verified code. (In my model it goes up to 100 and I get OOM finally after after a few hourse/days depending on the parameters of the model. If I skip loss.backward it is ok.)

**Comment by ericnakagawa** [comment]
I'm working to reproduce this now.

**Comment by peterszabo77** [comment]
There seems to be no leak if I use "torch.set_num_threads(1)" ... hope it helps

**Comment by rainmanwy** [comment]
Seems i also met such kind of problem.
When i using pytorch 1.0.0 (without cuda), the memory eated very fast.

After downgrade to pytorch 0.4.1.post2, memory looks very good:
              total        used        free      shared  buff/cache   available
Mem:          7.7Gi       2.1Gi       828Mi       324Mi       4.7Gi       4.9Gi
Swap:         4.0Gi       0.0Ki       4.0Gi

Below is my CNN:

```
        self.layer1 = nn.Sequential(
            nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1),
            nn.Dropout(0.3), 
            nn.ReLU(),
            nn.MaxPool2d(2))
        self.layer2 = nn.Sequential(
            nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1),
            nn.Dropout(0.3), 
            nn.ReLU(),
            nn.MaxPool2d(2))
        self.layer3 = nn.Sequential(
            nn.Conv2d(64, 64, kernel_size=3, stride=1, padding=1),
            nn.Dropout(0.3),  
            nn.ReLU(),
            nn.MaxPool2d(2))
        self.fc = nn.Sequential(
            nn.Linear((width // 8) * (height // 8) * 64, 1024),
            nn.Dropout(0.3), 
            nn.ReLU())
        self.rfc = nn.Sequential(
            nn.Linear(1024, charNum * ALL_CHAR_SET_LEN),
        )
```

**Comment by thomasj02** [comment]
I'm also seeing consistent increases in memory usage as training progresses, whether or not I do set_num_threads(1). @ericnakagawa were you able to reproduce this?

**Comment by gchanan** [comment]
@zou3519 is going to try to reproduce this.

**Comment by ezyang** [comment]
Note that there are some outstanding known memory leaks in autograd.  #24368 is the one that pops immediately into mind.

**Comment by zou3519** [comment]
tl;dr After investigation I determined that the reinforcement learning tutorial doesn't leak memory on pytorch 1.2. However, we are committed to making sure there are no more memory leaks so please create new issues with individual new examples of memory leaks and we will investigate with high priority.

I tried to repro this on my mac. I don't have a linux box with a monitor (which is a requirement to run the tutorial) but I suspect that the result should be the same.

My conclusion was that our reinforcement learning tutorial does NOT leak memory. However, it does grow in memory consumption. The reason why it grows in memory consumption is due to the model storing replay memory: https://github.com/pytorch/tutorials/blob/94cb6a3635b9b5ccb8a59c5346addba6901cd15d/intermediate_source/reinforcement_q_learning.py#L336

Once the replay memory limit is hit, the RAM consumption stays constant. This can be verified by setting the replay number to something low (i.e. 10) and watching RAM.

I know that this wasn't completely satisfactory because I'm sure that some of you have encountered memory leaks while training in PyTorch, but we are committed to fixing them so please submit additional bug reports about memory leaks for your individual examples and we can investigate them separately.
